pkg/ccl/backupccl/file_sst_sink.go

```diff
@@ -0,0 +1,329 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package backupccl
+
+import (
+	"context"
+	"fmt"
+	io "io"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/ccl/storageccl"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/builtins"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	hlc "github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/errors"
+	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/kr/pretty"
+)
+
+type sstSinkConf struct {
+	progCh   chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress
+	enc      *roachpb.FileEncryptionOptions
+	id       base.SQLInstanceID
+	settings *settings.Values
+}
+
+type fileSSTSink struct {
+	dest cloud.ExternalStorage
+	conf sstSinkConf
+
+	queue []returnedSpans
+	// queueCap is the maximum byte size that the queue can grow to.
+	queueCap int64
+	// queueSize is the current byte size of the queue.
+	queueSize int
+
+	sst     storage.SSTWriter
+	ctx     context.Context
+	cancel  func()
+	out     io.WriteCloser
+	outName string
+
+	flushedFiles    []BackupManifest_File
+	flushedSize     int64
+	flushedRevStart hlc.Timestamp
+	completedSpans  int32
+
+	stats struct {
+		files       int
+		flushes     int
+		oooFlushes  int
+		sizeFlushes int
+		spanGrows   int
+	}
+
+	memAcc struct {
+		ba            *mon.BoundAccount
+		reservedBytes int64
+	}
+}
+
+func makeFileSSTSink(
```

Comment:
I'm not sure that "fileSSTSink" is really telling me anything that "sstSink" didn't, but I also sympathize that it wasn't telling much / enough before.

I think the distinguishing characteristics of this thing are that it takes chunks of exported data (which just so happens to be in sst format), buffers and merges it, and then writes SSTables. So maybe bufferedSSTWriter or something like that?

Reply:
To me, calling the object a `FileSSTSink` highlights that it's not a generic sst sink. It outputs `BackupManifest_File` objects to the progress channel as one of two core outputs, and takes in a `BackupManifest_File` as an input in `exportedSpan` (though we're now more subtly calling that `metadata`). This isn't something that can be used to buffer any other kind of object, which took me a few hours to totally figure out.

Have I convinced?